samplers/aws/xray/remote_sampler.go

```diff
@@ -0,0 +1,189 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray // import "go.opentelemetry.io/contrib/samplers/aws/xray"
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/contrib/samplers/aws/xray/internal"
+	"go.opentelemetry.io/contrib/samplers/aws/xray/internal/util"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+
+	"github.com/go-logr/logr"
+)
+
+// remoteSampler is a sampler for AWS X-Ray which polls sampling rules and sampling targets
+// to make a sampling decision based on rules set by users on AWS X-Ray console
+type remoteSampler struct {
+	// manifest is the list of known centralized sampling rules.
+	manifest *internal.Manifest
+
+	// pollerStarted, if true represents rule and target pollers are started.
+	pollerStarted bool
+
+	// samplingRulesPollingInterval, default is 300 seconds.
+	samplingRulesPollingInterval time.Duration
+
+	// matching attribute
+	serviceName string
+
+	// matching attribute
+	cloudPlatform string
+
+	// fallback sampler
+	fallbackSampler *FallbackSampler
+
+	// logger for logging
+	logger logr.Logger
+}
+
+// Compile time assertion that remoteSampler implements the Sampler interface.
+var _ sdktrace.Sampler = (*remoteSampler)(nil)
+
+// NewRemoteSampler returns a sampler which decides to sample a given request or not
+// based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
+// sampling rules and sampling targets.
+func NewRemoteSampler(ctx context.Context, serviceName string, cloudPlatform string, opts ...Option) (sdktrace.Sampler, error) {
+	// create new config based on options or set to default values
+	cfg := newConfig(opts...)
+
+	// validate config
+	err := validateConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// create manifest with config
+	m, err := internal.NewManifest(cfg.endpoint, cfg.logger)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteSampler := &remoteSampler{
+		manifest:                     m,
+		samplingRulesPollingInterval: cfg.samplingRulesPollingInterval,
+		fallbackSampler:              NewFallbackSampler(),
+		serviceName:                  serviceName,
+		cloudPlatform:                cloudPlatform,
+		logger:                       cfg.logger,
+	}
+
+	// starts the rule and target poller
+	remoteSampler.start(ctx)
+
+	return remoteSampler, nil
+}
+
+// ShouldSample matches span attributes with retrieved sampling rules and perform sampling,
+// if rules does not match or manifest is expired then use fallback sampling.
+func (rs *remoteSampler) ShouldSample(parameters sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	if !rs.manifest.Expired() {
```

Comment:
```suggestion
	if rs.manifest.Expired() {
		// Use fallback sampler if manifest is expired
		rs.logger.V(5).Info("manifest is expired so using fallback sampling strategy")
		return rs.fallbackSampler.ShouldSample(parameters)
	}
```

Invert this condition and outdent the remainder, it'll be easier to follow.

Reply:
See the next comment.  The current code is doing double duty and should be split into two cases.  This immediate fallback if the manifest is expired and the later ultimate fallback if there are no rule matches.  They are separate cases and should be handled separately.  Doing so has the added benefit of making the logic flow clearer.